server/src/main/java/org/opensearch/action/pagination/WlmPaginationStrategy.java

```diff
@@ -0,0 +1,263 @@
+     * limited by pageSize. Sets responseToken if more results are available.
+     *
+     * @param rawStats unpaginated stats from all nodes
+     * @param requestedToken pagination token from the client, or null for the first page
+     * @param currentHash hash representing the current snapshot, used for token validation
+     * @return paginated sublist of WlmStats
+     */
+
+    private List<WlmStats> applyPagination(List<WlmStats> rawStats, WlmStrategyToken requestedToken, String currentHash) {
+        if (rawStats.isEmpty()) {
+            this.responseToken = null;
+            return Collections.emptyList();
+        }
+
+        List<WlmStats> perQueryGroupStats = extractQueryGroupStats(rawStats);
+
+        perQueryGroupStats = perQueryGroupStats.stream()
+                .sorted(sortOrder.apply(sortBy.getComparator()))
+                .collect(Collectors.toList());
+
+        int startIndex = getStartIndex(perQueryGroupStats, requestedToken);
+
+        List<WlmStats> page = getPage(perQueryGroupStats, startIndex);
+
+        setResponseToken(perQueryGroupStats, startIndex + page.size(), currentHash);
+
+        return page;
+    }
+
+    private List<WlmStats> extractQueryGroupStats(List<WlmStats> rawStats) {
+        List<WlmStats> result = new ArrayList<>();
+        for (WlmStats stat : rawStats) {
+            Map<String, WorkloadGroupStats.WorkloadGroupStatsHolder> queryGroups = stat.getWorkloadGroupStats().getStats();
+            for (Map.Entry<String, WorkloadGroupStats.WorkloadGroupStatsHolder> entry : queryGroups.entrySet()) {
+                String queryGroupId = entry.getKey();
+                WorkloadGroupStats singleQueryGroupStats = new WorkloadGroupStats(Map.of(queryGroupId, entry.getValue()));
+                result.add(new WlmStats(stat.getNode(), singleQueryGroupStats));
+            }
+        }
+        return result;
+    }
+
+    private int getStartIndex(List<WlmStats> sortedStats, WlmStrategyToken token) {
+        if (token == null) {
+            return 0;
+        }
+
+        OptionalInt index = findIndex(
+                sortedStats,
+                token.getNodeId(),
+                token.getQueryGroupId()
+        );
+
+        if (index.isEmpty()) {
+            throw new OpenSearchParseException("Invalid or outdated token: " + nextToken);
+        }
+
+        return index.getAsInt();
+    }
+
+    private List<WlmStats> getPage(List<WlmStats> stats, int startIndex) {
+        int endIndex = Math.min(startIndex + pageSize, stats.size());
+        return stats.subList(startIndex, endIndex);
+    }
+
+    private void setResponseToken(List<WlmStats> stats, int nextIndex, String currentHash) {
+        if (nextIndex < stats.size()) {
+            WlmStats lastEntry = stats.get(nextIndex - 1);
+            String nodeId = lastEntry.getNode().getId();
+            String queryGroupId = lastEntry.getWorkloadGroupStats().getStats().keySet().iterator().next();
```

Comment:
Should we also consider the SortOrder and Field in the token calculation, since it would help us set the sort order for the series of request/responses using the token